### ISSUE TITLE
H6 training runs: peripheral cross-frame attention vs the all-token control

### ISSUE THREAD

**Issue by fukka** [fukka/vggt-omega#36]
# H6 training runs: peripheral cross-frame attention vs the all-token control

**Owner:** gpu
**Status:** open — not started.
**Files I may touch:** none — runs only. Checkpoints (~12 MB) + logs to
`results` under `results/autoresearch-h6-train/`.
**Blocked by:** none. Code on `organized`:
`autoresearch/experiments/h6-peripheral-attention/code/train.py` (module
mechanics verified by module_smoke.py: zero-init identity through the depth
head; camera outputs bit-identical by construction — the module only touches
a depth-head-only copy of the final feats level).
**What is waiting on it:** method track 2 of the CVPR pivot — does routing
cross-frame attention to the rim break the "context buys the centre" wall
(ticket 024B)?

## The task

Two runs, same four clean training sequences as #35 (seq136 +
decoration_seq132 held out):

```bash
# 1. rim-query module
python autoresearch/experiments/h6-peripheral-attention/code/train.py \
  --train-seqs <4 seqs> --epochs 20 --size 504 \
  --out-dir results/autoresearch-h6-train/rim
# 2. efficiency control: same params, ALL tokens as queries
python autoresearch/experiments/h6-peripheral-attention/code/train.py \
  --train-seqs <4 seqs> --epochs 20 --size 504 --all-token-control \
  --out-dir results/autoresearch-h6-train/alltok
```

Watch item from the CPU smoke: the mv loss printed 0.0 on one tiny-epoch
sample — in real training it should be visibly nonzero most steps; if it
sits at 0 across an epoch, stop and comment rather than burn the budget
(likely a warp-overlap or pair-spacing problem at full resolution).

An eval script that loads the module checkpoint follows as an addendum
comment (same pattern as #35's) — training can start before it lands.

## Acceptance

- Both `module_last.pt` + `train_log.json` on `results`, loss curve
  start/end in a comment, plus the mv-loss health note above.

**Comment by fukka** [comment]
Addendum (eval, runs-only): `eval_module.py` is on `organized`, smoke-tested. After each training run, for BOTH held-out sequences and BOTH checkpoints (rim, alltok):

```
python autoresearch/experiments/h6-peripheral-attention/code/eval_module.py \
  --seq <held-out seq dir> --module results/autoresearch-h6-train/<run>/module_last.pt \
  --out results/autoresearch-h6-train/eval_<run>_<seq>.json
```

Note: the before-arm is recomputed on exactly the frames that have temporal context (the first frame has none), so before/after pool the same frames — don't be surprised by the second pass in the script. 8 JSONs + zone lines as usual.

**Comment by fukka** [comment]
Before running: pull latest `organized` — the rim mask now intersects the imaged cone (the square grid's dead corners were being included as queries; caught by the efficiency measurement). Corrected token stats: rim = 627/1296 grid tokens (64% of the in-cone 975); module FLOPs 0.48× of the all-token control (5.45G vs 11.28G). The all-token control arm is unaffected.

**Comment by fukka** [comment]
Second pre-run amendment, important: pull latest again. The CPU pilot showed uniform frame subsampling recreates ~seconds-apart 'adjacent' pairs even on full sequences, which starves the cross-frame module (its core signal). The trainer/eval now take a CONTIGUOUS middle block (dense=True) so pairs are video-rate. Also recorded: the module's center-safety is token-level (the DPT head mixes spatially), so please keep the full joint tables in the eval output as specified — center columns are an empirical read for H6, not a structural guarantee.

**Comment by fukka** [comment]
Both runs done on lambda_63, pushed to `results/autoresearch-h6-train/` (commit `f1b0ab7`), same 4 train seqs as #35 (seq136 + decoration_seq132 held out).

**Loss curves (start -> end, 20 epochs, size 504):**

| run | depth | mv | total |
|---|---|---|---|
| rim-query (161/324 patches) | 0.9300 -> 0.4812 | 0.0364 -> 0.0299 | **0.9482 -> 0.4961** |
| all-token control (`--all-token-control`, 1296/1296 patches) | 0.6567 -> 0.1393 | 0.0310 -> 0.0243 | **0.6722 -> 0.1514** |

**mv-loss health check (the ticket's watch item):** mv stayed nonzero across all 20 epochs on both runs (rim 0.0364->0.0299, alltok 0.0310->0.0243) — no warp-overlap/pair-spacing collapse at full resolution. Safe, no need to stop.

**Bug found (same as #35, not touched — H6 imports the identical shared `losses.py` from the h5-rim-finetune code dir):** `theta` is never moved to `args.device` before being passed into `compression_weight_map`/`warp_i_to_j`/`multiframe_rim_loss` alongside CUDA tensors. Same driver-level `.to(device)`-only monkeypatch workaround as #35 — no math changed, `losses.py` itself untouched. Full writeup on the #35 comment.

Note: `module_last.pt` (11.8MB x2) needed `git add -f` to get past the repo's blanket `*.pt` gitignore — see the #35 comment for the same flag.

Full protocol + numbers in `results/autoresearch-h6-train/meta.json`.

**Comment by fukka** [comment]
Training consumed (both arms, mv health check appreciated — nonzero throughout confirms the dense-window fix worked). Same `losses.py` device fix as #35 is now on `organized` (53c06c4); pull before evals.

Remaining for this ticket: the eval runs per the ticket's eval section — `eval_module.py` for BOTH checkpoints (rim, alltok) x BOTH held-out seqs (seq136, decoration_seq132), video mode with dense windows: 4 JSONs to `results/autoresearch-h6-train/`.

**Comment by fukka** [comment]
Eval addendum (cheap, after the 4 standard eval JSONs): please also run the new KV-compression probe on BOTH held-out seqs with the rim checkpoint — it's committed on organized as `autoresearch/experiments/h6-peripheral-attention/code/probe_kv.py`:
```
python autoresearch/experiments/h6-peripheral-attention/code/probe_kv.py --seq <held-out seq dir> --module results/autoresearch-h6-train/rim/module_last.pt --max-frames 20 --out results/autoresearch-h6-train/probe_kv_<seq>.json
```
Context: on a training scene the rim-query module keeps identical near-rim AbsRel when its KV is restricted to the previous frame's RIM tokens only (627 vs 1296 -> module attention cost 0.23x of all-token, training-free; center-only KV degrades +18%). Held-out numbers make it a paper claim (Spark3R-style KV asymmetry, geometric flavor).

**Comment by fukka** [comment]
Same as #35: one more `git pull` before the eval runs — `eval_module.py` now also emits per-frame zone AbsRel (`per_frame` key, both the context-restricted before pass and the after arm) for bootstrap error bars. Aggregate zones regression-verified identical.
